File: com.trollworks.gcs/src/com/trollworks/gcs/feature/ConditionalModifier.java
/*
 * Copyright ©1998-2021 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.feature;

import com.trollworks.gcs.datafile.DataFile;
import com.trollworks.gcs.ui.widget.outline.ListRow;
import com.trollworks.gcs.utility.json.JsonMap;
import com.trollworks.gcs.utility.json.JsonWriter;

import java.io.IOException;
import java.util.Map;
import java.util.Set;

public class ConditionalModifier extends Bonus {
    public static final String KEY_ROOT      = "conditional_modifier";
    public static final String KEY_SITUATION = "situation";

    private String mSituation;

    public ConditionalModifier() {
        super(1);
        mSituation = "triggering condition";
    }

    public ConditionalModifier(ConditionalModifier other) {
        super(other);
        mSituation = other.mSituation;
    }

    public ConditionalModifier(DataFile dataFile, JsonMap m) throws IOException {
        this();
        loadSelf(dataFile, m);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (obj instanceof ConditionalModifier && super.equals(obj)) {
            return mSituation.equals(((ConditionalModifier) obj).mSituation);
        }
        return false;
    }

    @Override
    public String getJSONTypeName() {
        return KEY_ROOT;
    }

    @Override
    public String getKey() {
        return KEY_ROOT;
    }

    @Override
    public Feature cloneFeature() {
        return new ConditionalModifier(this);
    }

    @Override
    protected void loadSelf(DataFile dataFile, JsonMap m) throws IOException {
        super.loadSelf(dataFile, m);
        setSituation(m.getString(KEY_SITUATION));
    }

    @Override
    protected void saveSelf(JsonWriter w) throws IOException {
        super.saveSelf(w);
        w.keyValue(KEY_SITUATION, mSituation);
    }

    public String getSituation() {
        return mSituation;
    }

    public void setSituation(String situation) {
        mSituation = situation;
    }

    @Override
    public void fillWithNameableKeys(Set<String> set) {
        ListRow.extractNameables(set, mSituation);
    }

    @Override
    public void applyNameableKeys(Map<String, String> map) {
        mSituation = ListRow.nameNameables(map, mSituation);
    }
}

File: model/gurps/feature/conditional_modifier.go
/*
 * Copyright ©1998-2022 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package feature

import (
	"fmt"

	"github.com/richardwilkes/gcs/v5/model/fxp"
	"github.com/richardwilkes/gcs/v5/model/gurps/nameables"
	"github.com/richardwilkes/toolbox/i18n"
	"github.com/richardwilkes/toolbox/xio"
)

var _ Bonus = &ConditionalModifier{}

// ConditionalModifier holds the data for a conditional modifier.
type ConditionalModifier struct {
	Type      Type         `json:"type"`
	Parent    fmt.Stringer `json:"-"`
	Situation string       `json:"situation,omitempty"`
	LeveledAmount
}

// NewConditionalModifierBonus creates a new ConditionalModifier.
func NewConditionalModifierBonus() *ConditionalModifier {
	return &ConditionalModifier{
		Type:          ConditionalModifierType,
		Situation:     i18n.Text("triggering condition"),
		LeveledAmount: LeveledAmount{Amount: fxp.One},
	}
}

// FeatureType implements Feature.
func (c *ConditionalModifier) FeatureType() Type {
	return c.Type
}

// Clone implements Feature.
func (c *ConditionalModifier) Clone() Feature {
	other := *c
	return &other
}

// FeatureMapKey implements Feature.
func (c *ConditionalModifier) FeatureMapKey() string {
	return ConditionalModifierType.Key()
}

// FillWithNameableKeys implements Feature.
func (c *ConditionalModifier) FillWithNameableKeys(m map[string]string) {
	nameables.Extract(c.Situation, m)
}

// ApplyNameableKeys implements Feature.
func (c *ConditionalModifier) ApplyNameableKeys(m map[string]string) {
	c.Situation = nameables.Apply(c.Situation, m)
}

// SetParent implements Bonus.
func (c *ConditionalModifier) SetParent(parent fmt.Stringer) {
	c.Parent = parent
}

// SetLevel implements Bonus.
func (c *ConditionalModifier) SetLevel(level fxp.Int) {
	c.Level = level
}

// AddToTooltip implements Bonus.
func (c *ConditionalModifier) AddToTooltip(buffer *xio.ByteBuffer) {
	basicAddToTooltip(c.Parent, &c.LeveledAmount, buffer)
}
